drivers/entropy/entropy_nrf5.c

```diff
@@ -8,92 +8,253 @@
+	} else {
+		u8_t *d, *s;
+		u8_t avail;
+
+		d = &rand[octets];
+		s = &rng->rand[first];
+
+		avail = rng->count - first;
+		if (octets < avail) {
+			remaining = avail + last - octets;
+			avail = octets;
+			first += avail;
+		} else {
+			remaining = last;
+			first = 0;
+		}
+
+		octets -= avail;
+
+		while (avail--) {
+			*(--d) = *s++;
+		}
+
+		if (octets && last) {
+			s = &rng->rand[0];
+
+			if (octets < last) {
+				remaining = last - octets;
+				last = octets;
+			} else {
+				remaining = 0;
+			}
+
+			first = last;
+			octets -= last;
+
+			while (last--) {
+				*(--d) = *s++;
+			}
+		}
+
+		rng->first = first;
 	}
-	nrf_rng_event_clear(NRF_RNG_EVENT_VALRDY);
 
-	/* Clear the Pending status of the RNG interrupt so that it could
-	 * wake up the core when the VALRDY event occurs again. */
-	NVIC_ClearPendingIRQ(RNG_IRQn);
+	if (remaining < rng->threshold) {
+		NRF_RNG->TASKS_START = 1;
+#if defined(CONFIG_BOARD_NRFXX_NWTSIM)
+		NRF_RNG_regw_sideeffects();
+#endif
+	}
 
-	return nrf_rng_random_value_get();
+	return octets;
 }
 
-static int entropy_nrf5_get_entropy(struct device *device, u8_t *buf, u16_t len)
+static inline int isr(struct rand *rng, bool store)
 {
-	/* Mark the peripheral as being used */
-	atomic_inc(&DEV_DATA(device)->user_count);
+	u8_t last;
+
+	if (!rng) {
+		return -ENOBUFS;
+	}
 
-	/* Disable the shortcut that stops the task after a byte is generated */
-	nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
+	last = rng->last + 1;
+	if (last == rng->count) {
+		last = 0;
+	}
 
-	/* Start the RNG generator peripheral */
-	nrf_rng_task_trigger(NRF_RNG_TASK_START);
+	if (last == rng->first) {
+		/* this condition should not happen, but due to probable race,
+		 * new value could be generated before NRF_RNG task is stopped.
+		 */
+		return -ENOBUFS;
+	}
 
-	while (len) {
-		*buf = entropy_nrf5_get_u8();
-		buf++;
-		len--;
+	if (!store) {
+		return -EBUSY;
 	}
 
-	/* Only stop the RNG generator peripheral if we're the last user */
-	if (atomic_dec(&DEV_DATA(device)->user_count) == 1) {
-		/* Disable the peripheral on the next VALRDY event */
-		nrf_rng_shorts_enable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
-
-		if (atomic_get(&DEV_DATA(device)->user_count) != 0) {
-			/* Race condition: another thread started to use
-			 * the peripheral while we were disabling it.
-			 * Enable the peripheral again
-			 */
-			nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
-			nrf_rng_task_trigger(NRF_RNG_TASK_START);
+	rng->rand[rng->last] = NRF_RNG->VALUE;
+	rng->last = last;
+
+	last = rng->last + 1;
+	if (last == rng->count) {
+		last = 0;
+	}
+
+	if (last == rng->first) {
+		return 0;
+	}
+
+	return -EBUSY;
+}
+
+static void isr_rand(void *arg)
+{
+	struct device *device = arg;
+
+	if (NRF_RNG->EVENTS_VALRDY) {
+		struct entropy_nrf5_dev_data *dev_data = DEV_DATA(device);
+		int ret;
+
+		ret = isr((struct rand *)dev_data->isr, true);
+		if (ret != -EBUSY) {
+			ret = isr((struct rand *)dev_data->thr,
```

Comment:
Why this call is made? Is this for flush random value in case of the race condition occurrence?

Reply:
^^I mean the second param value - `-ENOBUFS` might be returned in two cases.

Reply:
-ENOBUFS is returned to convey that a new rng value has not been put into ring buffer due to the ISR ring buffer being full (or no ISR ring buffer initialized). Hence, conveyed as "store" param to the next ring buffer (thread).

Same pattern of nested if-clause can be used to scale to any number of ring buffer population.

Reply:
Thanks. Then i understood this properly. Everything is right here.